notebooks/day2/flatland.jl

```diff
@@ -0,0 +1,875 @@
+
+# ╔═╡ ebf4a45a-6244-11eb-0965-197f536f8e87
+begin
+	area(shape::Shape) = missing
+end
+
+# ╔═╡ 230dd290-6303-11eb-0f55-311ef2b9541e
+
+
+# ╔═╡ 36cc0492-6246-11eb-38dd-4f42fb7066dc
+md"""
+## Moving, rotating and scaling
+
+Important: the functions work *in-place*, meaning that the modify your object (that is why use use `mutable` structures).
+
+For `Circle` and `AbstractRectangle` types, `rotate!` leaves them unchanged.
+
+**Note**: rotations are in radials, so between $0$ and $2\pi$.
+"""
+
+# ╔═╡ 317f43b8-6303-11eb-2d92-1b9457a68912
+
+
+# ╔═╡ 83c6d25c-6246-11eb-1a24-57e20f5e7262
+#=begin
+	function move!(shape::Shape, (dx, dx))
+		# move the shape
+		return shape
+	end
+end=#
+
+# ╔═╡ a1b2a4f8-6246-11eb-00ea-8f6042c72f4e
+#=begin
+	function rotate!(shape::Shape, dθ)
+		# rotate the shape, counterclockwise
+		return shape
+	end
+end=#
+
+# ╔═╡ b907e8fc-6246-11eb-0beb-bb44930d033c
+#=begin
+	function scale!(shape::Shape, a)
+		@assert a > 0 "scaling has to be a positive number"
+		# scale with a factor a
+		return shape
+	end
+end=#
+
+# ╔═╡ d6dd39be-6308-11eb-34bd-d16118d85051
+begin
+	
+	function move!(shape::Shape, (dx, dy))
+    shape.x += dx
+    shape.y += dy
+end
+
+function move!(shape::Triangle, (dx, dy))
+    shape.x1 += dx
+    shape.x2 += dx
+    shape.x3 += dx
+    shape.y1 += dy
+    shape.y2 += dy
+    shape.y3 += dy
+    shape
+end
+
+rotate!(shape::Union{Circle,AbstractRectangle}, dθ) = shape
+
+function rotate!(shape::RegularPolygon, dθ)
+    shape.θ += dθ
+    shape
+end
+
+
+function rotate!(shape::Triangle, dθ)
+    xc, yc = center(shape)
+    # center triangle
+    move!(shape, (-xc, -yc))
+    (x1, y1), (x2, y2), (x3, y3) = corners(shape)
+    cosdθ = cos(dθ)
+    sindθ = sin(dθ)
+    shape.x1 = x1 * cos(dθ) - y1 * sin(dθ)
+    shape.x2 = x2 * cos(dθ) - y2 * sin(dθ)
+    shape.x3 = x3 * cos(dθ) - y3 * sin(dθ)
+    shape.y1 = y1 * cos(dθ) + x1 * sin(dθ)
+    shape.y2 = y2 * cos(dθ) + x2 * sin(dθ)
+    shape.y3 = y3 * cos(dθ) + x3 * sin(dθ)
+    # set to original position
+    move!(shape, (xc, yc))
+    shape
+end
+
+function scale!(shape::Union{Circle,RegularPolygon}, a)
+    @assert a > 0 "scaling has to be a positive number"
+    shape.R *= a
+end
+
+function scale!(shape::Rectangle, a)
+    @assert a > 0 "scaling has to be a positive number"
+    shape.w *= a
+    shape.l *= a
+end
+
+function scale!(shape::Square, a)
+    @assert a > 0 "scaling has to be a positive number"
+    shape.w *= a
+end
+
+function scale!(shape::Triangle, a)
+    @assert a > 0 "scaling has to be a positive number"
+    xc, yc = center(shape)
+    move!(shape, (-xc, -yc))
+    shape.x1 *= a
+    shape.x2 *= a
+    shape.x3 *= a
+    shape.y1 *= a
+    shape.y2 *= a
+    shape.y3 *= a
+    move!(shape, (xc, yc))
+    return shape
+end
+	
+end
+
+# ╔═╡ d08ab6d0-6246-11eb-08a8-152f9802cdfc
+md"""
+## Plotting 
+
+OK, let's take a look at our shapes! We use `RecipesBase` to allow plotting.
```

Comment:
@DaanVanHauwermeiren  @MichielStock `RecipedBase` wordt hiervoor al uitgelegd zeker?

Reply:
Indien niet, uitleggen in de slides